numpy/lib/type_check.py

```diff
@@ -266,7 +272,19 @@ def iscomplexobj(x):
     True
 
     """
-    return issubclass(asarray(x).dtype.type, _nx.complexfloating)
+    try:
+        dtype = x.dtype
+    except AttributeError:
+        dtype = asarray(x).dtype
+        try:
+            return issubclass(dtype.type, _nx.complexfloating)
+        except AttributeError:
+            # The numpy.testing.utils.assert_equal routine relies on calling
+            # iscomplexobj with weird arguments (types) and catches ValueErrors
+            # only. We therefore convert AttributeError to ValueError to
+            # maintain a compatible interface.
```

Comment:
Is this really part of the public API? Could we simply fix `assert_equal` instead?

This seems like an unfortunate work around to preserve for backwards compatibility.